[mlir][dataflow] Fix for integer range analysis propagation bug

sabauma opened the pull request:
Integer range analysis will not update the range of an operation when any of the inferred input lattices are uninitialized. In the current behavior, all lattice values for non integer types are uninitialized.

For operations like arith.cmpf

```mlir
%3 = arith.cmpf ugt, %arg0, %arg1 : f32
```

that will result in the range of the output also being uninitialized, and so on for any consumer of the arith.cmpf result. When control-flow ops are involved, the lack of propagation results in incorrect ranges, as the back edges for loop carried values are not properly joined with the definitions from the body region.

For example, an scf.while loop whose body region produces a value that is in a dataflow relationship with some floating-point values through an arith.cmpf operation:

```mlir
func.func @test_bad_range(%arg0: f32, %arg1: f32) -> (index, index) {
  %c4 = arith.constant 4 : index
  %c1 = arith.constant 1 : index
  %c0 = arith.constant 0 : index

  %3 = arith.cmpf ugt, %arg0, %arg1 : f32

  %1:2 = scf.while (%arg2 = %c0, %arg3 = %c0) : (index, index) -> (index, index) {
    %2 = arith.cmpi ult, %arg2, %c4 : index
    scf.condition(%2) %arg2, %arg3 : index, index
  } do {
  ^bb0(%arg2: index, %arg3: index):
    %4 = arith.select %3, %arg3, %arg3 : index
    %5 = arith.addi %arg2, %c1 : index
    scf.yield %5, %4 : index, index
  }

  return %1#0, %1#1 : index, index
}
```

The existing behavior results in the control condition %2 being optimized to true, turning the while loop into an infinite loop. The update to %arg2 through the body region is never factored into the range calculation, as the ranges for the body ops all test as uninitialized.

This change causes all values initialized with setToEntryState to be set to some initialized range, even if the values are not integers.

Commit:
[mlir][dataflow] Fix for integer range analysis propagation bug

Integer range analysis will not update the range of an operation when
any of the inferred input lattices are uninitialized. In the current
behavior, all lattice values for non integer types are uninitialized.

For operations like arith.cmpf

```mlir
%3 = arith.cmpf ugt, %arg0, %arg1 : f32
```

that will result in the range of the output also being uninitialized,
and so on for any consumer of the arith.cmpf result. When control-flow
ops are involved, the lack of propagation results in incorrect ranges,
as the back edges for loop carried values are not properly joined with
the definitions from the body region.

For example, an scf.while loop whose body region produces a value that
is in a dataflow relationship with some floating-point values through
an arith.cmpf operation:

```mlir
func.func @test_bad_range(%arg0: f32, %arg1: f32) -> (index, index) {
  %c4 = arith.constant 4 : index
  %c1 = arith.constant 1 : index
  %c0 = arith.constant 0 : index

  %3 = arith.cmpf ugt, %arg0, %arg1 : f32

  %1:2 = scf.while (%arg2 = %c0, %arg3 = %c0) : (index, index) -> (index, index) {
    %2 = arith.cmpi ult, %arg2, %c4 : index
    scf.condition(%2) %arg2, %arg3 : index, index
  } do {
  ^bb0(%arg2: index, %arg3: index):
    %4 = arith.select %3, %arg3, %arg3 : index
    %5 = arith.addi %arg2, %c1 : index
    scf.yield %5, %4 : index, index
  }

  return %1#0, %1#1 : index, index
}
```

The existing behavior results in the control condition %2 being
optimized to true, turning the while loop into an infinite loop. The
update to %arg2 through the body region is never factored into the range
calculation, as the ranges for the body ops all test as uninitialized.

This change causes all values initialized with setToEntryState to
be set to some initialized range, even if the values are not integers.

ubfx commented:
I agree that the control flow issue you have laid out is a problem. But to me it seems like assigning "fake" ranges to non-Integer values might lead to other problems down the line. 

Maybe we should think about this check? Why don't we do `setAllToEntryStates()` on the Operation's results when any of the operands are unitialized?
https://github.com/llvm/llvm-project/blob/main/mlir/lib/Analysis/DataFlow/IntegerRangeAnalysis.cpp#L76

krzysz00 commented:
Re that check, I *think* the point was to prevent having to create fake constant ranges for non-integer inputs.

I think the right solution would be to change the interface from `ArrayRef<ConstantIntRanges>` to something like `ArrayRef<optional<ConstantIntRanges>>` 

Or we could fake things by, for example, declaring non-integers to be zero-width, for instance.

sabauma commented:
> Re that check, I _think_ the point was to prevent having to create fake constant ranges for non-integer inputs.

@krzysz00, yes That check was added explicitly to prevent a crash relating to faking the integer ranges. That problem seems to no longer occur.

> Maybe we should think about this check? Why don't we do setAllToEntryStates() on the Operation's results when any of the operands are unitialized?

@ubfx I agree my proposed solution is kind of odd. As @krzysz00 mentioned in his comment, I suspect removing that check is the right approach, but it would require changes to the `InferIntRangeInterface`. I would be happy to take a stab at that, but it would be a more invasive change.

ubfx commented:
Ok, then I'm +1 on changing the interface

sabauma committed:
Rework integer range analysis interfaces

Modify the integer range analysis interfaces to handle uninitialized
values by allowing the inferred input ranges to be optional.

krzysz00 commented on mlir/lib/Analysis/DataFlow/IntegerRangeAnalysis.cpp:
```diff
@@ -36,10 +36,26 @@
 
+namespace {
+
+OptionalIntRanges getOptionalRange(const IntegerValueRange &range) {
```
Nit: I think this'd usually be a `static` and not `namespace`d

sabauma replied:
Removed these functions

krzysz00 reviewed (commented):
I'm not going to LGTM this so late at night, but this seems like a reasonable fix.

(Having an "intrusive" None value is the other option we could think of here ,but I can't see a way to do that to the range type, so ... yeah, optional is what it'll probably be)

This might need an announcement somewhere?

ubfx commented on mlir/lib/Dialect/Arith/IR/InferIntRangeInterfaceImpls.cpp:
```diff
@@ -46,243 +45,271 @@ void arith::ConstantOp::inferResultRanges(ArrayRef<ConstantIntRanges> argRanges,
                                       SetIntRangeFn setResultRange) {
-  setResultRange(getResult(), inferMul(argRanges, convertArithOverflowFlags(
-                                                      getOverflowFlags())));
+  auto infer = inferFromOptionals([this](ArrayRef<ConstantIntRanges> ranges) {
```
Could passing a lambda as `function_ref` be a  problem? I think I tried something like this and it caused crashes on some compilers/platforms, maybe this is the reason for the MSVC build failure?

sabauma replied:
I'll try changing this to `std::function`. Given how this function returns a lambda which captures the `function_ref`, I think there is a good chance this is related to the crashes.

sabauma commented on mlir/include/mlir/Interfaces/Utils/InferIntRangeCommon.h:
```diff
@@ -27,6 +27,9 @@ namespace intrange {
     function_ref<ConstantIntRanges(ArrayRef<ConstantIntRanges>)>;
 
+using OptionalRangeFn =
+    std::function<OptionalIntRanges(ArrayRef<OptionalIntRanges>)>;
```
I was a uncomfortable using `function-ref` since some functions return `OptionalRangeFn`.

sabauma replied:
Given the way some of the new utility functions are manipulating lambdas, the `InferRangeFn` also needs to be converted to `std::function`.

joker-eph replied:
`function_ref` is unsafe as a return value in general, but it should always be safe when transiently used as function arguments. 
It's slightly more murky when a reference is kept, because the caller must keep the called object alive (but that's just "business as usual": anytime you pass a pointer/reference to a method that keeps  a copy of it you have this situation).

It can be tricky in situations like this:

```
std::function<void()> bar();
std::function<void()> foo(function_ref<void()> callable) {
  // The returned std::function copies the callable function_ref
  return [=]() { 
     callable();
  };
```

If we write:
```
std::function<void()> bar_result = bar();
std::function<void()> foo_result = foo(bar_result);
```

Then it's just fine (but careful: foo_result has a reference to `bar_result`, now bar_result must outlive uses of `foo_result`).
So it is buggy to write:

```
std::function<void()> foo_result = foo(bar());
```

Because now `foo_result` has a reference to a dangling object (the result of `bar()`).

sabauma commented:
> (Having an "intrusive" None value is the other option we could think of here ,but I can't see a way to do that to the range type, so ... yeah, optional is what it'll probably be)

@krzysz00 Another option I was considering was to rework the interfaces to take `IntegerRangeValue`. I think that might clarify the API of the interfaces a little.

Commit:
Convert uses of OptionalIntRange to IntegerValueRange

IntegerValueRange already exists and encodes the extact information that
we want to represent with OptionalIntRange. This makes the APIs clearer
than passing an std::optional everywhere.

krzysz00 reviewed (commented):
So to keep the discussion going, I remember keeping `ConstantIntRanges` and the lattices separate back in the day so as to make the integer range inference usable for cases other than that one dataflow analysis. But that's not a super hard goal and so this might be a reasonable solution.

Re std::function vs function_ref, I don't have a strong opinion there - all I can say is that the intent was to pass in a transient pointer to a callback. Seems reasonable to make that change.

But also, wrt the interface and backwards-compatibily, how about
1. Provide a new method, possibly with the same name (or if not, something like `inferResultRangesFromOptional`) that takes `ArrayRef<std::optional<ConstantIntRanges>>` instead of `ArrayRef<ConstantIntRanges>`
2. Implement the optional version in terms of the non-optional version by default: that is, if someone doesn't implement `inferResultRanges(ArrayRef<std::optional<ConstantIntRanges>>, SetResultFn)`, it'll loop through the inputs and, if they're all not `none`, call `inferResultRanges()` on the unwrapped values. This gives everyone an upgrade path.

Commit:
Add new IntegerRangeAnalysis interface method

This new method allows downstream implementers to easily opt into the
old behavior while providing an easy way to transition to the more
powerful interface methods.

sabauma commented:
@krzysz00 I've applied the suggested changes. Extending the interface resulted in a much cleaner design and I was able to revert most of the surgery done to the dialects using the `InferIntRangeInterface`.

> Provide a new method, possibly with the same name (or if not, something like inferResultRangesFromOptional) that takes ArrayRef<std::optional<ConstantIntRanges>> instead of ArrayRef<ConstantIntRanges>

I was not able to use the same name overloaded on different input types. This seems to be a limitation of the interface system.

krzysz00 reviewed (approved):
I'm glad we finally managed to find a good approach here! Thanks for putting up with the endless rounds of revisions.

Approved!
